### ISSUE TITLE
Test on own dataset / No instances to display

### ISSUE THREAD

**Issue by nayzen** [matterport/Mask_RCNN#912]
Hi everyone,

I'm currently using this repo for an object detection problem with a lot of classes.
I trained on my own dataset (37 classes) and I used the splash balloon effect which works but I would like bounding boxes and scores confidence to see which object is detected. 
That is why I have tried to test with the "demo.ipynb" code, to load my weights and to test the detection on my own dataset.
There is no error but I got this message : **"NO INSTANCE TO DISPLAY"**

I am a beginner in the computer vision field but I think I'm close to have results.
Here is my curves of my training :

![loss_curve](https://user-images.githubusercontent.com/37196095/45153316-ccfabb80-b1d3-11e8-9856-f6201edfd28e.JPG)

![val_loss](https://user-images.githubusercontent.com/37196095/45153360-fa476980-b1d3-11e8-81e4-9b941bf60d08.JPG)

But I think there is a problem when the weights are loaded. 
In the original "demo.ipynb", **I have an error with these lines which load the weights:**

`model = modellib.MaskRCNN(mode="inference", model_dir=MODEL_DIR, config=config)
model.load_weights(COCO_MODEL_PATH, by_name=True)`

**My error disappears when I replace with this line:**

`model.load_weights(COCO_MODEL_PATH, by_name=True, exclude=[ "mrcnn_class_logits", "mrcnn_bbox_fc", "mrcnn_bbox", "mrcnn_mask"])`

But **"NO INSTANCE TO DISPLAY"**

Maybe somebody can help me with this ?

Regards,
Antoine

**Comment by fastlater** [comment]
Better check the nucleus demo. There, you will see many ways to use the methods from visualize.py

**Comment by camillemontalcini** [comment]
Hey! I have been experiencing this **NO INSTANCE TO DISPLAY**, but actually I notice that I had the   DETECTION_MIN_CONFIDENCE  parameter (in config.py) that was equal to 0.7, hence all the prediction below this threshold was not displayed. Puting this to 0.5 (hence removing the effect in case of binary classification) allowed to display all instances. Dont know if this helps.

**Comment by nayzen** [comment]
@fastlater Thanks for your answer ! Are you talking about this repo:
**https://github.com/wanwanbeen/maskrcnn_nuclei**   ?

@camillemontalcini Thanks for having shared your feedbacks ! I tried to change the value to 0.5 and it displays instances but the detections are not good at all :'(

(I forgot to mention that I have used VGG Image Annotator tool for my groundtruth data, and I don't know if the .json output is different in structure from the COCO .json annotations...)

**Comment by fastlater** [comment]
@camillemontalcini  https://github.com/matterport/Mask_RCNN/tree/master/samples/nucleus

**Comment by AliceDinh** [comment]
You said: there is a problem when the weights are loaded. What's exact problem?

**Comment by nayzen** [comment]
@AliceDinh Here is my jupyter notebook code (the demo.ipynb) modified for my needs. I have 37 classes and I load the weights "**mask_rcnn_f100_0150.h5**". Firstly, I test the detection on images of the training set "**imagesf100**".

![screen1](https://user-images.githubusercontent.com/37196095/45289895-8d431500-b4ee-11e8-9a0b-b8318de7d645.PNG)

Then, here is my config, and I load my weights **which are not trained on MS-COCO dataset**.

![screen2](https://user-images.githubusercontent.com/37196095/45289896-8d431500-b4ee-11e8-9c6e-79fd32074675.PNG)

Here I have my 37 classes but I don't know if it is necessary like COCO.

![screen3](https://user-images.githubusercontent.com/37196095/45289897-8ddbab80-b4ee-11e8-9229-ee889a34631e.PNG)

Here is my detection and "No instances to display" even if I have a DETECTION_MIN_CONFIDENCE of 0.7 or 0.5.

![screen4](https://user-images.githubusercontent.com/37196095/45289898-8ddbab80-b4ee-11e8-886c-9a143937d101.PNG)

Am I wrong somewhere ?

**Comment by AliceDinh** [comment]
@nayzen I reckoned that there is nothing wrong with the notebook. However, probably your annotation, what took did you use to create the annotation (VIA?) the format of annotation? and your code F100.py, you should have look these again.

**Comment by nayzen** [comment]
I used the VIA tool to do my annotations. The format is .json and this is the organisation:

![organisation](https://user-images.githubusercontent.com/37196095/45342426-3a289b00-b59e-11e8-9667-12e08ded1615.png)

For the code "F100.py", I listened to your advices by email.

**Comment by AliceDinh** [comment]
I think you have to try with lower confidence to see any instance, you used 0.5, let's try lower. And try with the images in the training set. That's all I can suggest. Sorry can't help much

**Comment by nayzen** [comment]
Thanks anyway @AliceDinh , for your help :-)
When I test the "demo.ipynb" with **balloon weights**, I see instances but the detection are not good at all:
![bad_detection2](https://user-images.githubusercontent.com/37196095/45360016-9a373580-b5ce-11e8-9bd8-5825b7c307d7.PNG)
![bad_detection3](https://user-images.githubusercontent.com/37196095/45360018-9acfcc00-b5ce-11e8-9980-c5a597d6cf31.PNG)

**I don't understand why. All I can say is that I must exclude some layers otherwise it doesn't work :(**

![second_solution](https://user-images.githubusercontent.com/37196095/45360518-18e0a280-b5d0-11e8-9528-c0d6e5b9ffd8.png)

![model1](https://user-images.githubusercontent.com/37196095/45360423-dae37e80-b5cf-11e8-99e5-eb248e24c8e7.PNG)

Here is the error : "**_ValueError: Dimension 1 in both shapes must be equal, but are 8 and 324. Shapes are [1024,8] and [1024,324]. for 'Assign_1366' (op: 'Assign') with input shapes: [1024,8], [1024,324]._**"

If someone could explain to me.. I would be grateful.

**Comment by nayzen** [comment]
I fixed my problem deleting the .h5 file at the root, and copying it again.

![good_detection1](https://user-images.githubusercontent.com/37196095/45362022-049ea480-b5d4-11e8-92af-9bd807d772a5.png)
![good_detection2](https://user-images.githubusercontent.com/37196095/45362023-049ea480-b5d4-11e8-9b3c-bba4cdb96ea1.png)
![good_detection3](https://user-images.githubusercontent.com/37196095/45362025-049ea480-b5d4-11e8-98cb-3c47177bf3c3.png)
![good_detection4](https://user-images.githubusercontent.com/37196095/45362026-05373b00-b5d4-11e8-9b6e-ca56289ef852.png)

**Comment by greebear** [comment]
@nayzen 
Hello! 

I met the same question. Only can I get a good result when I put
`exclude=[ "mrcnn_class_logits", "mrcnn_bbox_fc", "mrcnn_bbox", "mrcnn_mask"]`
in the code.
u said: I fixed my problem deleting the .h5 file at the root, and copying it again.
do u mean that u found a way to get a good result without code  like 
`exclude=[ "mrcnn_class_logits", "mrcnn_bbox_fc", "mrcnn_bbox", "mrcnn_mask"]`

thx for reading

**Comment by nayzen** [comment]
#greebear

PS: And don't forget to include the background for the first class name ! ('BG', 'balloon', '...',...)

**Comment by BeyondCloud** [comment]
I've solved this problem by modifying utils.py line 866 from
`shift = np.array([0, 0, 1, 1])`
to
`shift = np.array([0, 0, 1., 1.])`
It's simply a data type error which will cause float number becomes zero, leading to divide by zero error

**Comment by lapetite123** [comment]
how can i get the loss and val loss curve?

**Comment by nayzen** [comment]
@lapetite123  Hello,
You can visualize during training (or after) the loss and val curve opening a prompt command and hit:
tensorboard -logdir=C:\.....\Mask_RCNN-master\logs\folder-containing-.h5-file

**Comment by ganleiboy** [comment]
> I've solved this problem by modifying utils.py line 866 from
> `shift = np.array([0, 0, 1, 1])`
> to
> `shift = np.array([0, 0, 1., 1.])`
> It's simply a data type error which will cause float number becomes zero, leading to divide by zero error

Thank you very much !

**Comment by 739982423** [comment]
> I think you have to try with lower confidence to see any instance, you used 0.5, let's try lower. And try with the images in the training set. That's all I can suggest. Sorry can't help much

it works, thanks!

**Comment by AherSiddharth** [comment]
> Hey! I have been experiencing this **NO INSTANCE TO DISPLAY**, but actually I notice that I had the DETECTION_MIN_CONFIDENCE parameter (in config.py) that was equal to 0.7, hence all the prediction below this threshold was not displayed. Puting this to 0.5 allowed to display more instances for example. Dont know if this helps.

Thank you!
